# Overlap with Methylation

The methylation peaks will be read in from the right files. 

```{r use_dir}
data_dir <- "/mlab/data/rmflight/Documents/projects/work/fondufe-mittendorf_lab/parp1_data"
library(GenomicRanges)
library(magrittr)
options(mc.cores = 10)
library(parallel)
library(ggplot2)
```

Based on previous manuscripts, it may be a good idea to restrict both the methylation and parp1 reads to a 2kb window surrounding annotated transcription start sites (tss). 

```{r read_tss}
tss_file <- file.path(data_dir, "ensGene_TTSS.csv")
tss_data <- read.table(tss_file, header = TRUE, sep = ",", stringsAsFactors = FALSE)
tss_regions <- GRanges(seqnames = tss_data$chrom,
                       strand = tss_data$strand,
                       ranges = IRanges(start = tss_data$txStart,
                                        end = tss_data$txEnd),
                       names = tss_data$name)
rm(tss_data)
```

Generate 2kb regions around each *tss*.

```{r windowed_tss}
tss_windows <- GRanges(seqnames = seqnames(tss_regions),
                       strand = strand(tss_regions),
                       ranges = IRanges(start = start(tss_regions) - 1000,
                                        end = start(tss_regions) + 1000),
                       names = names(tss_regions))
```


```{r read_methylation}
methyl_files <- file.path(data_dir, dir(data_dir, pattern = "CTCF_narrow_peak_H"))
methyl_names <- strsplit(methyl_files, "_") %>% sapply(., function(x){x[6]})
methyl_names <- strtrim(methyl_names, nchar(methyl_names) - 4)

names(methyl_files) <- methyl_names

methyl_data <- lapply(methyl_files, function(x){
  tmp <- read.table(x, header = TRUE, sep = ",")
  names(tmp) <- c("chrom", "start", "end", "name", "score", "strand", "signal", "pvalue", "qvalue", "other", "id")
  GRanges(seqnames = tmp$chrom,
          ranges = IRanges(start = tmp$start, end = tmp$end),
          mcols = DataFrame(tmp[, c("score", "signal", "pvalue")]), overlap = 0, overlap_ratio = 0)
})
methyl_data <- GRangesList(methyl_data)
```

We will also store the total number of reads from *LN4* and *LN5* samples.

```{r}
total_reads <- c(ln4 = 60486202, ln5 = 54125517)
```


Now lets go through the *parp1* nucleosome data. For each chromosome, we will subset those reads that are within the methylation reads.

```{r get_chromosome_within_methyl}
get_chr <- function(filename){
  chr_part <- strsplit(filename, "_")[[1]][5]
  substr(chr_part, 1, nchar(chr_part)-4)
}

ln4_files <- file.path(data_dir, dir(data_dir, pattern = "YFM_LN4_chr"))

ln4_overlap <- mclapply(ln4_files, function(x){
  tmp <- read.table(x, header = TRUE, sep = ",")
  tmpRange <- GRanges(seqnames = get_chr(x),
                      ranges = IRanges(start = tmp$startx, width = 1))
  lapply(methyl_data, function(y){
    countOverlaps(y, tmpRange)
  })
})

names(ln4_overlap) <- sapply(ln4_files, get_chr)
```

Before we get the data out, lets go through for a given methylation status, and check that each chromosome gives different non-zero entries.

```{r check_non_zero}
check_indices <- lapply(ln4_overlap, function(x){
  which(x[[1]] > 0)
})
check_indices <- unlist(check_indices)
sum(duplicated(check_indices))
```

So now we will add the number of overlapped data in `ln4_overlapped` to the data in `methyl_data`. We already created this slot, so hopefully we will avoid any memory problems.

```{r copy_overlap_counts}
methyl_names <- names(methyl_data)
ln4_names <- names(ln4_overlap)

for (i_methyl in methyl_names){
  for (i_ln4 in ln4_names){
    use_indices <- which(ln4_overlap[[i_ln4]][[i_methyl]] > 0)
    mcols(methyl_data[[i_methyl]])$overlap[use_indices] <- ln4_overlap[[i_ln4]][[i_methyl]][use_indices]
    mcols(methyl_data[[i_methyl]])$overlap_ratio <- mcols(methyl_data[[i_methyl]])$overlap / total_reads["ln4"]
  }
}
```


Now with the counts in place, we can graph the data and see what we have. Because we have so many points, we will use a sub-sample of 1000 points to see what exactly is going on.

```{r graphit}
use_vars <- c("mcols.signal", "overlap_ratio")
subsamples <- lapply(methyl_names, function(x){
  tmp_data <- subsample_nonzeros(mcols(methyl_data[[x]]), c("mcols.signal", "overlap_ratio"), non_zero = "both", n_points = 1000)
  ggplot(tmp_data, aes(x = mcols.signal, y = overlap_ratio)) + geom_point() + ggtitle(x) + scale_x_log10() + scale_y_log10()
})
```

hmmm, correlations are rather poor. 