# Overlap with Methylation

The methylation peaks will be read in from the right files. 

```{r use_dir}
data_dir <- "/mlab/data/rmflight/Documents/projects/work/fondufe-mittendorf_lab/parp1_data"
library(GenomicRanges)
library(magrittr)
options(mc.cores = 10)
library(parallel)
```

```{r read_methylation}
methyl_files <- file.path(data_dir, dir(data_dir, pattern = "CTCF_narrow_peak_H"))
methyl_names <- strsplit(methyl_files, "_") %>% sapply(., function(x){x[6]})
methyl_names <- strtrim(methyl_names, nchar(methyl_names) - 4)

names(methyl_files) <- methyl_names

methyl_data <- lapply(methyl_files, function(x){
  tmp <- read.table(x, header = TRUE, sep = ",")
  names(tmp) <- c("chrom", "start", "end", "name", "score", "strand", "signal", "pvalue", "qvalue", "other", "id")
  GRanges(seqnames = tmp$chrom,
          ranges = IRanges(start = tmp$start, end = tmp$end),
          mcols = DataFrame(tmp[, c("score", "signal", "pvalue")]))
})
methyl_data <- GRangesList(methyl_data)
```


Now lets go through the *parp1* nucleosome data. For each chromosome, we will subset those reads that are within the methylation reads.

```{r get_chromosome_within_methyl}
get_chr <- function(filename){
  chr_part <- strsplit(filename, "_")[[1]][5]
  substr(chr_part, 1, nchar(chr_part)-4)
}

ln4_files <- file.path(data_dir, dir(data_dir, pattern = "YFM_LN4_chr"))

ln4_overlap <- mclapply(ln4_files, function(x){
  tmp <- read.table(x, header = TRUE, sep = ",")
  tmpRange <- GRanges(seqnames = get_chr(x),
                      ranges = IRanges(start = tmp$startx, width = 1))
  lapply(methyl_data, function(y){
    countOverlaps(y, tmpRange)
  })
})

names(ln4_overlap) <- sapply(ln4_files, get_chr)

```

Now with methylation in